File: go/WorkQueue.go
// Package workqueue is the Go implementation of a work queue, on top of a redis database, with implementations in Python, Rust, Go, Node.js (TypeScript) and Dotnet (C#).
//
// For an overview of how the work queue works, it’s limitations, and the general concepts and
// implementations in other languages, please read the redis-work-queue readme at
// https://github.com/MeVitae/redis-work-queue/blob/main/README.md.
//
// # Setup
//
//	import (
//	    "github.com/redis/go-redis/v9"
//	    workqueue "github.com/mevitae/redis-work-queue/go"
//	)
//
//	db := redis.NewClient(&redis.Options{
//	    Addr: "your-redis-server:1234",
//	})
//	workQueue := workqueue.NewWorkQueue(workqueue.KeyPrefix("example_work_queue"))
//
// # Adding Items
//
//	// From bytes:
//	exampleItem := NewItem([]byte("[1,2,3]"))
//	// Or from JSON:
//	jsonItem, err := NewItemFromJSONData([1, 2, 3])
//	// ...
//	err = workQueue.AddItem(ctx, db, jsonItem)
//
// # Completing work
//
// Please read the documentation on leasing and completing items at
// https://github.com/MeVitae/redis-work-queue/blob/main/README.md#leasing-an-item
//
//	for (;;) {
//	   job, err := workQueue.Lease(ctx, db, true, 0, 5*time.Second)
//	   if err != nil { panic(err) }
//	   doSomeWork(job)
//	   _, err = workQueue.Complete(ctx, db, job)
//	   if err != nil { panic(err) }
//	}
//
// # Handling errors
//
// Please read the documentation on handling errors at
// https://github.com/MeVitae/redis-work-queue/blob/main/README.md#handling-errors
//
//	for (;;) {
//	   job, err := workQueue.Lease(ctx, db, true, 0, 5*time.Second)
//	   if err != nil { panic(err) }
//	   err = doSomeWork(job)
//	   if err == nil {
//	       // Mark successful jobs as complete
//	       _, err = workQueue.Complete(ctx, db, job)
//	       if err != nil { panic(err) }
//	   } else if !shouldRetryAfter(err) {
//	       // Errors that shouldn't cause a retry should mark the job as complete so it isn't tried
//	       // again.
//	       logError(err)
//	       _, err = workQueue.Complete(ctx, db, job)
//	       if err != nil { panic(err) }
//	   } else {
//	       // Drop a job that should be retried - it will be returned to the work queue after the
//	       // (5 second) lease expires.
//	   }
//	}
package workqueue

import (
	"context"
	"fmt"
	"slices"
	"time"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

const never time.Duration = 0

// WorkQueue backed by a redis database.
type WorkQueue struct {
	// session is a unique ID for this instance
	session string
	// mainQueueKey is the key for the list of items in the queue
	mainQueueKey string
	/// processingKey is the key for the list of items being processed
	processingKey string
	// leaseKey is the  key prefix for lease entries
	leaseKey KeyPrefix
	// itemDataKey is the key prefix for item data
	itemDataKey KeyPrefix
}

func NewWorkQueue(name KeyPrefix) WorkQueue {
	return WorkQueue{
		session:       uuid.NewString(),
		mainQueueKey:  name.Of(":queue"),
		processingKey: name.Of(":processing"),
		leaseKey:      name.Concat(":lease:"),
		itemDataKey:   name.Concat(":item:"),
	}
}

// AddItem to the work queue.
//
// If an item with the same ID already exists, this item is not added, and false is returned. Otherwise, if the item is added true is returned.
//
// If you know the item ID is unique, and not already in the queue, use the optimised WorkQueue.AddUniqueItem instead.
func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item Item) (bool, error) {
	added, err := db.SetNX(ctx, workQueue.itemDataKey.Of(item.ID), item.Data, never).Result()
	if added {
		err = db.LPush(ctx, workQueue.mainQueueKey, item.ID).Err()
	}
	return added, err
}

// AddItemToPipeline adds an item, which is known to have an ID not already in the queue, to the work queue. This adds the redis commands onto the pipeline passed.
//
// Use [WorkQueue.AddUniqueItem] if you don't want to pass a pipeline directly.
func (workQueue *WorkQueue) AddUniqueItemToPipeline(ctx context.Context, pipeline redis.Pipeliner, item Item) {
	// Add the item data
	// NOTE: it's important that the data is added first, otherwise someone could pop the item
	// before the data is ready
	pipeline.Set(ctx, workQueue.itemDataKey.Of(item.ID), item.Data, never)
	// Then add the id to the work queue
	pipeline.LPush(ctx, workQueue.mainQueueKey, item.ID)
}

// AddItem, which is known to have an ID not already in the queue, to the work queue.
//
// This creates a pipeline and executes it on the database.
func (workQueue *WorkQueue) AddUniqueItem(ctx context.Context, db *redis.Client, item Item) error {
	pipeline := db.Pipeline()
	workQueue.AddUniqueItemToPipeline(ctx, pipeline, item)
	_, err := pipeline.Exec(ctx)
	return err
}

// Return the length of the work queue (not including items being processed, see
// [WorkQueue.Processing]).
func (workQueue *WorkQueue) QueueLen(ctx context.Context, db *redis.Client) (int64, error) {
	return db.LLen(ctx, workQueue.mainQueueKey).Result()
}

// Processing returns the number of items being processed.
func (workQueue *WorkQueue) Processing(ctx context.Context, db *redis.Client) (int64, error) {
	return db.LLen(ctx, workQueue.processingKey).Result()
}

// Request a work lease the work queue. This should be called by a worker to get work to complete.
// When completed, the [WorkQueue.Complete] method should be called.
//
// If block is true, the function will return either when a job is leased or after timeout if
// timeout isn't 0.
//
// If the job is not completed before the end of lease_duration, another worker may pick up the same
// job. It is not a problem if a job is marked as done more than once.
//
// If no job is available before the timeout, (nil, nil) is returned.
//
// If you've not already done it, it's worth reading the documentation on leasing items at
// https://github.com/MeVitae/redis-work-queue/blob/main/README.md#leasing-an-item
func (workQueue *WorkQueue) Lease(
	ctx context.Context,
	db *redis.Client,
	block bool,
	timeout time.Duration,
	leaseDuration time.Duration,
) (*Item, error) {
	for {
		// First, to get an item, we try to move an item from the main queue to the processing list.
		var command *redis.StringCmd
		if block {
			command = db.BRPopLPush(ctx, workQueue.mainQueueKey, workQueue.processingKey, timeout)
		} else {
			command = db.RPopLPush(ctx, workQueue.mainQueueKey, workQueue.processingKey)
		}
		itemId, err := command.Result()
		if itemId == "" || err != nil {
			// A nil error indicates no job available
			if err == redis.Nil {
				return nil, nil
			}
			return nil, err
		}

		// Get the item's data
		data, err := db.Get(ctx, workQueue.itemDataKey.Of(itemId)).Bytes()
		if err != nil {
			if err == redis.Nil {
				if block && timeout == 0 {
					continue
				}
				return nil, nil
			}
			return nil, err
		}

		// Now setup the lease item.
		// NOTE: Racing for a lease is ok
		err = db.SetEx(ctx, workQueue.leaseKey.Of(itemId), workQueue.session, leaseDuration).Err()

		return &Item{
			ID:   itemId,
			Data: data,
		}, err
	}
}

func (workQueue *WorkQueue) leaseExists(ctx context.Context, db *redis.Client, itemId string) (bool, error) {
	exists, err := db.Exists(ctx, workQueue.itemDataKey.Of(itemId)).Result()
	return exists > 0, err
}

func (workQueue *WorkQueue) LightClean(ctx context.Context, db *redis.Client) error {
	// A light clean only checks items in the processing queue
	itemIds, err := db.LRange(ctx, workQueue.processingKey, 0, -1).Result()
	if err != nil {
		return fmt.Errorf("failed to list processing queue: %w", err)
	}
	for _, itemId := range itemIds {
		leaseExists, err := workQueue.leaseExists(ctx, db, itemId)
		if err != nil {
			return fmt.Errorf("failed to check if lease exists for %s: %w", itemId, err)
		}
		// If there's no lease for the item, then it should be reset.
		if !leaseExists {
			// We also check that the item actually exists before pushing it back to the main queue
			itemExists, err := db.Exists(ctx, workQueue.itemDataKey.Of(itemId)).Result()
			if err != nil {
				return fmt.Errorf("failed to check if item %s exists: %w", itemId, err)
			}
			if itemExists != 0 {
				fmt.Println(itemId, "has no lease, it will be reset")
				pipeline := db.Pipeline()
				pipeline.LRem(ctx, workQueue.processingKey, 0, itemId)
				pipeline.LPush(ctx, workQueue.mainQueueKey, itemId)
				_, err := pipeline.Exec(ctx)
				if err != nil {
					return fmt.Errorf("failed to move %s from processing queue to main queue: %w", itemId, err)
				}
			} else {
				fmt.Println(itemId, "was in the processing queue but does not exist")
				err = db.LRem(ctx, workQueue.processingKey, 0, itemId).Err()
				if err != nil {
					return fmt.Errorf("failed to remove %s from processing queue: %w", itemId, err)
				}
			}
		}
	}
	return nil
}

func (workQueue *WorkQueue) DeepClean(ctx context.Context, db *redis.Client) error {
	// A deep clean checks all data keys
	pipeline := db.Pipeline()
	itemDataKeys := db.Keys(ctx, workQueue.itemDataKey.Of("*"))
	mainQueueRes := db.LRange(ctx, workQueue.mainQueueKey, 0, -1)
	_, err := pipeline.Exec(ctx)
	if err != nil {
		return fmt.Errorf("failed to list item data keys: %w", err)
	}
	mainQueue := mainQueueRes.Val()
	for _, itemDataKey := range itemDataKeys.Val() {
		itemId := itemDataKey[:len(workQueue.itemDataKey)]
		leaseExists, err := workQueue.leaseExists(ctx, db, itemId)
		if err != nil {
			return fmt.Errorf("failed to check if lease exists for %s: %w", itemId, err)
		}
		// If the item isn't in the queue, and there's no lease for the item, then it should be reset.
		if !slices.Contains(mainQueue, itemId) && !leaseExists {
			fmt.Println(itemId, "has no lease, it will be reset")
			pipeline := db.Pipeline()
			pipeline.LRem(ctx, workQueue.processingKey, 0, itemId)
			pipeline.LPush(ctx, workQueue.mainQueueKey, itemId)
			_, err := pipeline.Exec(ctx)
			if err != nil {
				return fmt.Errorf("failed to move %s from processing queue to main queue: %w", itemId, err)
			}
		}
	}
	return nil
}

// Complete marks a job as completed and remove it from the work queue. After Complete has been
// called (and returns true), no workers will receive this job again.
//
// Complete returns a boolean indicating if *the job has been removed* **and** *this worker was the
// first worker to call Complete*. So, while lease might give the same job to multiple workers,
// complete will return true for only one worker.
func (workQueue *WorkQueue) Complete(ctx context.Context, db *redis.Client, item *Item) (bool, error) {
	pipeline := db.Pipeline()
	delResult := pipeline.Del(ctx, workQueue.itemDataKey.Of(item.ID))
	pipeline.LRem(ctx, workQueue.processingKey, 0, item.ID)
	pipeline.Del(ctx, workQueue.leaseKey.Of(item.ID))
	_, err := pipeline.Exec(ctx)
	return delResult.Val() != 0, err
}

File: node/src/WorkQueue.ts
/**
 * @module WorkQueue
 * @description A work queue backed by a redis database.
 */

import Redis, {ChainableCommander} from 'ioredis';
import {v4 as uuidv4} from 'uuid';
import {Item} from './Item';
import {KeyPrefix} from './KeyPrefix';

export {KeyPrefix, Item};

/**
 * A work queue backed by a redis database.
 */
export class WorkQueue {
  private session: string;
  private mainQueueKey: string;
  private processingKey: string;
  private leaseKey: KeyPrefix;
  private itemDataKey: KeyPrefix;

  /**
   * @param {KeyPrefix} name This is the prefix created for the WorkQueue.
   */
  constructor(name: KeyPrefix) {
    this.mainQueueKey = name.of(':queue');
    this.processingKey = name.of(':processing');
    this.session = uuidv4();
    this.leaseKey = name.concat(':lease:');
    this.itemDataKey = name.concat(':item:');
  }

  /**
   * Add an item to the work queue.
   *
   * If an item with the same ID already exists, this item is not added, and `false` is returned.
   * Otherwise, if the item is added `true` is returned.
   *
   * If you know the item ID is unique, and not already in the queue, use the optimised
   * `WorkQueue.add_unique_item` instead.
   *
   * @param {Redis} db - The Redis Connection.
   * @param {Item} item - The item to be added..
   * @returns {Promise<bool>} - A boolean indicating if the item was added.
   */
  async addItem(db: Redis, item: Item): Promise<boolean> {
    const added = await db.setnx(this.itemDataKey.of(item.id), item.data) > 0;
    if (added) await db.lpush(this.mainQueueKey, item.id);
    return added;
  }

  /**
   * Add an item, which is known to have an ID not already in the queue, to the work queue. This
   * adds the redis commands onto the pipeline passed.
   * 
   * Use `WorkQueue.add_unique_item` if you don't want to pass a pipeline directly.
   *
   * @param {Pipeline} pipeline - The pipeline that the data will be executed.
   * @param {Item} item - The Item to be added.
   */
  addUniqueItemToPipeline(pipeline: ChainableCommander, item: Item) {
    // NOTE: it's important that the data is added first, otherwise someone before the data is ready.
    pipeline.set(this.itemDataKey.of(item.id), item.data);
    // Then add the id to the work queue
    pipeline.lpush(this.mainQueueKey, item.id);
  }

  /**
   * Add an item, which is known to have an ID not already in the queue, to the work queue.
   *
   * This creates a pipeline and executes it on the database.
   *
   * @param {Redis} db - The Redis Connection.
   * @param {Item} item - The item to be added.
   */
  async addUniqueItem(db: Redis, item: Item): Promise<void> {
    const pipeline = db.pipeline();
    this.addUniqueItemToPipeline(pipeline, item);
    await pipeline.exec();
  }

  /**
   * Return the length of the work queue (not including items being processed, see
   * `WorkQueue.processing`).
   *
   * @param {Redis} db - The Redis Connection.
   * @returns {Promise<number>} The length of the work queue (not including items being processed).
   */
  queueLen(db: Redis): Promise<number> {
    return db.llen(this.mainQueueKey);
  }

  /**
   * Return the number of items being processed.
   *
   * @param {Redis} db - The Redis Connection.
   * @returns {Promise<number>} The number of items being processed.
   */
  processing(db: Redis): Promise<number> {
    return db.llen(this.processingKey);
  }

  /**
   * Request a work lease from the work queue. This should be called by a worker to get work to
   * complete.  When completed, the `complete` method should be called.
   *
   * If `block` is true, the function will return either when a job is leased or after `timeout`
   * seconds if `timeout` isn't 0.
   *
   * If the job is not completed before the end of `leaseDuration`, another worker may pick up the same job.
   *
   * It is not a problem if a job is marked as `done` more than once.
   *
   * If you haven't already, it's worth reading the documentation on leasing items:
   * https://github.com/MeVitae/redis-work-queue/blob/main/README.md#leasing-an-item
   *
   * @param {Redis} db - The Redis Connection.
   * @param {number} leaseSecs - The number of seconds that the lease should hold.
   * @param {boolean} block - If false, the method will return immediately if there is no item.
   * @param {number} timeout - The number of seconds the lease will time out at.
   * @returns {Promise<Item>} The item to process!
   */
  async lease(
    db: Redis,
    leaseSecs: number,
    block = true,
    timeout = 0,
  ): Promise<Item | null> {
    for (;;) {
      let maybeItemId: string | null = null;

      // Try to move an item from the main queue to the processing list.
      if (block) {
        maybeItemId = await db.brpoplpush(
          this.mainQueueKey,
          this.processingKey,
          timeout
        );
      } else {
        maybeItemId = await db.rpoplpush(this.mainQueueKey, this.processingKey);
      }

      if (maybeItemId == null) {
        return null;
      }

      const itemId = maybeItemId;

      let data: Buffer | null = await db.getBuffer(this.itemDataKey.of(itemId));

      if (data == null) {
        if (block && timeout === 0) continue;
        return null;
      }

      // Setup the lease item.
      await db.setex(this.leaseKey.of(itemId), leaseSecs, this.session);
      return new Item(data, itemId);
    }
  }

  /**
   * Marks a job as completed and remove it from the work queue. After `complete` has been called
   * (and returns `true`), no workers will receive this job again.
   *
   * @param {Redis} db - The Redis connection.
   * @param {Item} item - The Item to complete.
   * @returns {boolean} a boolean indicating if *the job has been removed* **and** *this worker was the first worker to call `complete`*. So, while lease might give the same job to multiple workers, complete will return `true` for only one worker.
   */
  async complete(db: Redis, item: Item): Promise<boolean> {
    const results = await db.pipeline()
        .del(this.itemDataKey.of(item.id))
        .lrem(this.processingKey, 0, item.id)
        .del(this.leaseKey.of(item.id))
        .exec();
    return Boolean(results && results[0][1]);
  }

  /**
   * Check if a lease exists for an `itemId`.
   *
   * @param {Redis} db - The Redis Connection.
   * @param {string} itemId
   * @returns {Promise<boolean>} A boolean indicating if a lease exists for the item.
   */
  private async leaseExists(db: Redis, itemId: string): Promise<boolean> {
    const exists = await db.exists(this.leaseKey.of(itemId));
    return exists > 0;
  }

  /**
   * Check if an item exists with `itemId`.
   *
   * @param {Redis} db - The Redis Connection.
   * @param {string} itemId
   * @returns {Promise<boolean>} A boolean indicating if the item exists.
   */
  private async itemExists(db: Redis, itemId: string): Promise<boolean> {
    const exists = await db.exists(this.leaseKey.of(itemId));
    return exists > 0;
  }

  async lightClean(db: Redis) {
    const processing: Array<string> = await db.lrange(
      this.processingKey,
      0, -1,
    );
    for (const itemId of processing) {
      // If there's no lease for the item, then it should be reset.
      if (!(await this.leaseExists(db, itemId))) {
        // We also check that the item actually exists before pushing it back to the main queue
        if (await this.itemExists(db, itemId)) {
          console.log(itemId, "has no lease, it will be reset");
          await db.pipeline()
            .lrem(this.processingKey, 0, itemId)
            .lpush(this.mainQueueKey, itemId)
            .exec();
        } else {
            console.log(itemId, "was in the processing queue but does not exist");
            await db.lrem(this.processingKey, 0, itemId);
        }
      }
    }
  }

  async deepClean(db: Redis) {
    // A deep clean checks all data keys
    const itemResults = await db.pipeline()
      .keys(this.itemDataKey.of("*"))
      .lrange(this.mainQueueKey, 0, -1)
      .exec();
    if (!itemResults) throw new Error("pipeline did not return results");
    const [[_err_a, itemDataKeys], [_err_b, mainQueueUntyped]] = itemResults;
    const mainQueue = mainQueueUntyped as string[];
    for (const itemDataKey of itemDataKeys as string[]) {
      const itemId = itemDataKey.slice(this.itemDataKey.prefix.length);
      // If the item isn't in the queue, and there's no lease for the item, then it should be reset.
      if (!mainQueue.includes(itemId) && !(await this.leaseExists(db, itemId))) {
        console.log(itemId, "has no lease, it will be reset");
        await db.pipeline()
          .lrem(this.processingKey, 0, itemId)
          .lpush(this.mainQueueKey, itemId)
          .exec();
      }
    }
  }
}
